concepts/infrastructure-reconciliation/README.md

```diff
@@ -1,41 +1,172 @@
+
+### Issuing RBAC-based kubeconfigs
+
+To provide an environment, where certain Operators are not able to leverage the cluster access, it is required to open up a possibility for the constrained kubeconfigs to be created. This ensures that Operators can work only with the resources they require access to and security-wise, potential malicious parties won't be able to use kubeconfigs issued for the Operators to execute cluster-wide attacks.
+
+The solution is to introduce yet another operator which would be designed around creating Roles, RoleBindings and ServiceAccounts in the Kubernetes cluster. Those resources can be used to issue limited kubeconfigs, which will be rotated in a time-based manner. Creation of those resources in Kyma Runtimes will be done using the master kubeconfigs coming from the Gardener, which will not be stored anywhere in the Control Plane.
+
+### Dependency management
+
+It should be the responsibility of the Component Reconcilers to figure out if all dependencies exist and the reconciliation can be completed. This is true for all the infrastructure-related artefacts, such as the cluster state and kubeconfigs. If dependencies are not ready, the Component Reconciler should return the information that it is not ready and wait for the next trigger (time or event).
+
+## High-level architecture
 
 ![](assets/infrastructure-reconciliation.svg)
 
-Mothership Reconciler should provide common functionality related to watching changes and triggering Component Reconcilers in response to events or with the time-based schedule. Such aspects as queueing reconciliation requests, consolidating them, updating reconciliation status, retries and back-off strategy should be implemented once (in the Mothership). 
+The three new operators would need to be introduced:
+- Cluster Operator
+- Subscription Operator
+- Kubeconfig Operator
+
+Each of those components has a clear set of responsibilities that together handle the flow that creates a new, complete, Kyma-ready  Kubernetes cluster.
+
+Along the operators, there are three new Custom Resources, that serve as an API for the components:
+- Cluster CR
+- SubscriptionRequest CR
+- KubeconfigRequest CR
+
+### Cluster Operator and the Cluster CR
+
+The Cluster Operator's main responsibility is to manage the Cluster CRs, request subscriptions for the provisioning and create Shoot CRs in the Gardener cluster.
+
+On top of that, there is also a requirement to provide out-of-the-box admin roles to some users, this is the only place where the Cluster Operator interacts directly with the created cluster.
+
+Also, the exemplary Cluster CR resource looks as follows:
```

Comment:
QQ: Is there a document somewhere explaining why we do not use Kubernetes Cluster SIG? I would have imagined you could simply write a cluster class: https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass.html

Reply:
I've read a bit into that and I find the resources described in the Kubernetes Cluster SIG more infrastructure-focused.

Here, we want to somehow unite both the infrastructure and business aspects of the cluster provisioning. We already have a Gardener's Shoot CRs for the low-level infrastructure configuration, I would rather like to abstract from that and provide a simple Cluster API, while hiding the network/worker config underneath - we can contain that inside the Cluster Manager's logic.

Anyways, after few meetings the idea is to use some of the patterns that Kubernets Cluster SIG utilizes when it comes to the K8S resources and how they are tied to eachother.